Fix `__Pyx_PyBool_FromLong` for statements with internal commas.

Currently, Cython introduces a compiler error in the C++ compiler when converting a C/C++ `bint/bool` to a Python bool in a complex statement, due to `__Pyx_PyBool_FromLong` being a macro.

# Reproducing the Error

**header.h**
```cpp
template <class T1, class T2>
class BinaryAnd {
public:
    static T1 call(T1 x, T2 y) { return x & y; }
};
```

**source.pyx**
```cython
cdef extern from "header.h":
    cdef cppclass BinaryAnd[T1, T2]:
        @staticmethod
        T1 call(T1 x, T2 y)

def return_bool(bool x, bool y):
    return BinaryAnd[bool, bool].call(x, y)
```

# Error

```bash
error: macro "__Pyx_PyBool_FromLong" passed 2 arguments, but takes just 1
   __pyx_t_1 = __Pyx_PyBool_FromLong(BinaryAnd<bool,bool> ::call(__pyx_v_x, __pyx_v_y)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 189, __pyx_L1_error)
                                                                                      ^
cpp_templates.cpp: In function ‘PyObject* __pyx_pf_13cpp_templates_20test_compound_bool_return(PyObject*, bool, bool)’:
cpp_templates.cpp:3238:15: error: ‘__Pyx_PyBool_FromLong’ was not declared in this scope
   __pyx_t_1 = __Pyx_PyBool_FromLong(BinaryAnd<bool,bool> ::call(__pyx_v_x, __pyx_v_y)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 189, __pyx_L1_error)
               ^~~~~~~~~~~~~~~~~~~~~
cpp_templates.cpp:3238:15: note: suggested alternative: ‘__Pyx_PyInt_From_long’
   __pyx_t_1 = __Pyx_PyBool_FromLong(BinaryAnd<bool,bool> ::call(__pyx_v_x, __pyx_v_y)); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 189, __pyx_L1_error)
               ^~~~~~~~~~~~~~~~~~~~~
               __Pyx_PyInt_From_long
```

# Solution

The fix is simple: convert the macro to an inlined function.

```diff
diff --git a/Cython/Utility/TypeConversion.c b/Cython/Utility/TypeConversion.c
index 734e293ed..20af1d92a 100644
--- a/Cython/Utility/TypeConversion.c
+++ b/Cython/Utility/TypeConversion.c
@@ -84,7 +84,8 @@ static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u) {
 
 #define __Pyx_NewRef(obj) (Py_INCREF(obj), obj)
 #define __Pyx_Owned_Py_None(b) __Pyx_NewRef(Py_None)
-#define __Pyx_PyBool_FromLong(b) ((b) ? __Pyx_NewRef(Py_True) : __Pyx_NewRef(Py_False))
+static CYTHON_INLINE PyObject * __Pyx_PyBool_FromLong(long b);
 static CYTHON_INLINE int __Pyx_PyObject_IsTrue(PyObject*);
 static CYTHON_INLINE PyObject* __Pyx_PyNumber_IntOrLong(PyObject* x);
 
@@ -402,6 +403,12 @@ static CYTHON_INLINE Py_ssize_t __Pyx_PyIndex_AsSsize_t(PyObject* b) {
   return ival;
 }
 
+
+static CYTHON_INLINE PyObject * __Pyx_PyBool_FromLong(long b) {
+  return b ? __Pyx_NewRef(Py_True) : __Pyx_NewRef(Py_False);
+}
+
+
 static CYTHON_INLINE PyObject * __Pyx_PyInt_FromSize_t(size_t ival) {
     return PyInt_FromSize_t(ival);
 }
```

--- a/Cython/Utility/TypeConversion.c
+++ b/Cython/Utility/TypeConversion.c
@@ -84,7 +84,7 @@ static CYTHON_INLINE size_t __Pyx_Py_UNICODE_strlen(const Py_UNICODE *u) {
 
 #define __Pyx_NewRef(obj) (Py_INCREF(obj), obj)
 #define __Pyx_Owned_Py_None(b) __Pyx_NewRef(Py_None)
-#define __Pyx_PyBool_FromLong(b) ((b) ? __Pyx_NewRef(Py_True) : __Pyx_NewRef(Py_False))
+static CYTHON_INLINE PyObject * __Pyx_PyBool_FromLong(long b);
 static CYTHON_INLINE int __Pyx_PyObject_IsTrue(PyObject*);
 static CYTHON_INLINE PyObject* __Pyx_PyNumber_IntOrLong(PyObject* x);
 
@@ -401,6 +401,12 @@ static CYTHON_INLINE Py_ssize_t __Pyx_PyIndex_AsSsize_t(PyObject* b) {
   Py_DECREF(x);
   return ival;
 }
+
+
+static CYTHON_INLINE PyObject * __Pyx_PyBool_FromLong(long b) {
+  return b ? __Pyx_NewRef(Py_True) : __Pyx_NewRef(Py_False);
+}
+
 
 static CYTHON_INLINE PyObject * __Pyx_PyInt_FromSize_t(size_t ival) {
     return PyInt_FromSize_t(ival);

--- a/tests/run/cpp_bool_template_return.pyx
+++ b/tests/run/cpp_bool_template_return.pyx
@@ -1,0 +1,16 @@
+# tag: cpp
+
+from libcpp cimport bool
+
+cdef extern from "cpp_templates_helper.h":
+    cdef cppclass BinaryAnd[T1, T2]:
+        @staticmethod
+        T1 call(T1 x, T2 y)
+
+
+def test_compound_bool_return(bool x, bool y):
+    """
+    >>> test_compound_bool_return(True, False)
+    False
+    """
+    return BinaryAnd[bool, bool].call(x, y)

--- a/tests/run/cpp_templates_helper.h
+++ b/tests/run/cpp_templates_helper.h
@@ -44,3 +44,9 @@ class Div {
 public:
     static T half(T value) { return value / 2; }
 };
+
+template <class T1, class T2>
+class BinaryAnd {
+public:
+    static T1 call(T1 x, T2 y) { return x & y; }
+};

